### PR TITLE
Added struct for pet, modified contract functions

### LINKED ISSUE #1: Suggestions for a better implementation
A struct can be used instead of just an index to display more information about the pet like type of animal, age, etc.

### PR DESCRIPTION
**Pull request by ShivRaiGithub**
Fixes #1
Added a new struct of Pet to better define the kind of pet, modified functions accordingly, added a new function to get unadopted pets, made some other modifications for better security and gas optimizations.

### CODE REVIEW

**Comment by ShivRaiGithub** [comment]
The contract was tested on remix for quick test of all functions. For hardhat test, a new pr can be made.

File: contracts/PetAdoption.sol [hunk 1]
```diff
@@ -1,42 +1,103 @@
...
-    address public owner;
-    uint public petIndex = 0;
-    uint[] public allAdoptedPets;
+    // Struct to store pet details
+    struct Pet {
+        uint256 id;
+        string category;
+        string breed;
+    }
+    address private immutable i_owner;
```
Comment (swrno) [inline contracts/PetAdoption.sol:11]: why you are making address private?

File: contracts/PetAdoption.sol [hunk 2]
```diff
@@ -1,42 +1,103 @@
...
-    mapping(address => uint[]) public ownerToPets;
+    Pet[] private allPets;
+    mapping(uint256 => address) private petIdToOwner;
+    mapping(address => uint256[]) private ownerToPetIds;
 
-    constructor(uint initialPetIndex){
-        owner = msg.sender;
-        petIndex = initialPetIndex; 
+    constructor(){
+        i_owner = msg.sender;
```
Comment (swrno) [inline contracts/PetAdoption.sol:20]: why you are changing the name to `i_owner`?

**Review by swrno — COMMENTED** [review]
Overall did it worked with the reach app? if yes attach a screenshot or video of the UI.

**Comment by ShivRaiGithub** [comment]
> Overall did it worked with the reach app? if yes attach a screenshot or video of the UI.

I haven't integrated the changes with the frontend yet, I would if this pr is accepted because since it is dealing with the contract itself, any change would have to be implemented for the frontend again.
The frontend is still able to render though.

File: contracts/PetAdoption.sol (hunk 1 above)
Comment (ShivRaiGithub) [reply contracts/PetAdoption.sol:11]: We already have a function to get the address of owner, as such there is no requirement to leave it public. Also private and immutable are a little more gas efficient.

File: contracts/PetAdoption.sol (hunk 2 above)
Comment (ShivRaiGithub) [reply contracts/PetAdoption.sol:20]: Just a standard practice. It shows that owner is immutable ( hence "i_").

**Comment by swrno** [comment]
> > Overall did it worked with the reach app? if yes attach a screenshot or video of the UI.
> 
> I haven't integrated the changes with the frontend yet, I would if this pr is accepted because since it is dealing with the contract itself, any change would have to be implemented for the frontend again.
> The frontend is still able to render though.


Well, I don't think so. I think the contract and the frontend should be in sync, so do the changes which are required. Don't take stress about the labels and all, these things are for learning only.

**Comment by ShivRaiGithub** [comment]
Ok sure, will create a new PR with the frontend integrated too. However, please take care of the labels since it was told by a mentor in wob discord that contributions will be counted through labels. Thank you.

**Comment by swrno** [comment]
> Ok sure, will create a new PR with the frontend integrated too. However, please take care of the labels since it was told by a mentor in wob discord that contributions will be counted through labels. Thank you. 

Sorry, I can't merge any PR before I ensure it end to end working. Otherwise the project will go unstructured, I don't want that.

**Comment by ShivRaiGithub** [comment]
> > Ok sure, will create a new PR with the frontend integrated too. However, please take care of the labels since it was told by a mentor in wob discord that contributions will be counted through labels. Thank you.
> 
> Sorry, I can't merge any PR before I ensure it end to end working. Otherwise the project will go unstructured, I don't want that.

Yes, I meant a new pr including the frontend and contract changes. This one can be closed.

File: contracts/PetAdoption.sol (hunk 1 above)
Comment (swrno) [reply contracts/PetAdoption.sol:11]: We don't want the owner to be private.

**Pull request closed without merging**